nucypher/blockchain/eth/sol/compile.py

```diff
@@ -16,157 +16,301 @@
 """
 
 
-import collections
-from os.path import abspath, dirname
-
-import itertools
 import os
 import re
+from pathlib import Path
 from twisted.logger import Logger
-from typing import List, NamedTuple, Optional, Set
+from typing import Dict, Iterator, List, Optional, Pattern, Tuple, TypeVar, Union
 
-from nucypher.blockchain.eth.sol import SOLIDITY_COMPILER_VERSION
+from nucypher.blockchain.eth.sol import SOLIDITY_COMPILER_VERSION as SOURCE_VERSION
+from nucypher.exceptions import DevelopmentInstallationRequired
 
+LOG: Logger = Logger('solidity-compiler')
 
-class SourceDirs(NamedTuple):
-    root_source_dir: str
-    other_source_dirs: Optional[Set[str]] = None
 
+#
+# Types
+#
 
-class SolidityCompiler:
+# TODO: Is there a better way to specify complex dictionaries with variable keys?
+CompiledContracts = TypeVar('CompiledContracts', bound=Dict[str, Dict[str, List[Dict[str, Union[str, List[Dict[str, str]]]]]]])
```

Comment:
May be more readable if classes are used instead of anonymous dicts.

Reply:
Good Idea :+1: